File: coding/src/com/algos/li14_linkedlist/_L876_MiddleOfTheLinkedList.java
package  com.algos.li14_linkedlist;

import  com.algos.li30_model.ListNode;

public class _L876_MiddleOfTheLinkedList {

	public static void main(String[] args) {
		_L876_MiddleOfTheLinkedList obj = new _L876_MiddleOfTheLinkedList();
		_G01_SingleLinkedList list = new _G01_SingleLinkedList();
		list.insertAtEnd(1);
		list.insertAtEnd(2);
		list.insertAtEnd(3);
		list.insertAtEnd(4);
		list.insertAtEnd(5);
		
		System.out.println(obj.middleNode(list.head));
	}

	// Problem: https://leetcode.com/problems/middle-of-the-linked-list/
	// Idea: Use fast and slow pointers
	public ListNode middleNode(ListNode head) {
        ListNode slow = head,fast = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return slow;
    }
}

File: dsa/topics/09_linkedlist/python/l876_middle_of_the_linked_list.py
"""
Problem: https://leetcode.com/problems/middle-of-the-linked-list/

Idea:
Use fast and slow pointers
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def middle_node(head: ListNode) -> ListNode:
    slow = fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    return slow


# Example usage
if __name__ == "__main__":
    # 1 -> 2 -> 3 -> 4 -> 5
    vals = [1, 2, 3, 4, 5]
    head = None
    curr = None

    for v in vals:
        node = ListNode(v)
        if not head:
            head = curr = node
        else:
            curr.next = node
            curr = curr.next

    mid = middle_node(head)
    print(mid.val)  # 3
